Source/WebCore/html/HTMLAttachmentElement.cpp

```diff
@@ -146,28 +175,74 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
+
+    updateSave(attributeWithoutSynchronization(saveQualifiedName()));
+}
 
-    m_elementWithAction = HTMLDivElement::create(document());
-    m_elementWithAction->setIdAttribute(attachmentActionIdentifier());
-    if (const auto& action = attachmentActionForDisplay(); !action.isEmpty())
-        m_elementWithAction->setInnerText(String { action });
-    container->appendChild(*m_elementWithAction);
+class AttachmentSaveEventListener final : public EventListener {
+public:
+    static Ref<AttachmentSaveEventListener> create(HTMLAttachmentElement& attachment, const AtomString& eventType) { return adoptRef(*new AttachmentSaveEventListener(attachment, eventType)); }
 
-    m_elementWithTitle = HTMLDivElement::create(document());
-    m_elementWithTitle->setIdAttribute(attachmentTitleIdentifier());
-    if (auto title = attachmentTitleForDisplay(); !title.isEmpty())
-        m_elementWithTitle->setInnerText(WTFMove(title));
-    container->appendChild(*m_elementWithTitle);
+    bool operator==(const EventListener& other) const final
+    {
+        return this == &other;
+    }
 
-    m_elementWithSubtitle = HTMLDivElement::create(document());
-    m_elementWithSubtitle->setIdAttribute(attachmentSubtitleIdentifier());
-    if (auto subtitle = attachmentSubtitleForDisplay(); !subtitle.isEmpty())
-        m_elementWithSubtitle->setInnerText(WTFMove(subtitle));
-    container->appendChild(*m_elementWithSubtitle);
+    void handleEvent(ScriptExecutionContext&, Event& event) final
+    {
+        if (event.type() == eventNames().clickEvent) {
+            event.preventDefault();
+            event.stopPropagation();
+            event.stopImmediatePropagation();
+
+            CustomEvent::Init init;
+            init.bubbles = true;
+            init.cancelable = true;
+            init.composed = true;
+            m_attachment->dispatchEvent(CustomEvent::create(m_eventType, init, event.isTrusted() ? Event::IsTrusted::Yes : Event::IsTrusted::No));
+        } else
+            ASSERT_NOT_REACHED();
+    }
+
+private:
+    explicit AttachmentSaveEventListener(HTMLAttachmentElement& attachment, AtomString eventType)
+        : EventListener(CPPEventListenerType)
+        , m_attachment(attachment)
+        , m_eventType(eventType)
+    {
+    }
+
+    WeakPtr<HTMLAttachmentElement, WeakPtrImplWithEventTargetData> m_attachment;
+    AtomString m_eventType;
+};
+
+void HTMLAttachmentElement::updateSave(const AtomString& eventTypeName)
+{
+    if (!m_elementContainer)
+        return;
+
+    if (eventTypeName.isEmpty()) {
+        if (m_elementWithSave) {
+            m_elementContainer->removeChild(*m_elementWithSave);
+            m_elementWithSave = nullptr;
+        }
+        return;
+    }
+
+    if (!m_elementWithSave)
+        m_elementWithSave = createShadowDivElement(attachmentSaveIdentifier(), String { });
+
+    m_elementWithSave->addEventListener(eventNames().clickEvent, AttachmentSaveEventListener::create(*this, eventTypeName), { });
```

Comment:
So if we change multiple times the "save" attribute, do we end up having multiple event listeners?

Reply:
Oh, whoops, good catch! 😅

Reply:
You can just have a single type of `AttachmentSaveEventListener` that reads off the save attribute fwiw, instead of passing in the save attribute to the event listener. That should remove the need to remove/re-add the listener.